---
title: "Draw karyotype plot on any organism"
author: "Sichong Peng"
date: '2019-02-13'
draft: yes
image:
  caption: ''
  focal_point: ''
slug: draw-karyotype-plot-on-any-organism
tags:
- visualization
- plot
categories:
- R
- Visualization
---



<p>In my work, we often need to visualize the location of a set of variants in a genome. When you have a dozen variants, it’s usually not too bad to go through them manually and place them on chromosomes. However, as the number of variants grow, it gets annoying very quickly!</p>
<p>Like most of you working on non-model organisms, I have long been jealousy of human geneticists’ ability to quickly draw a fancy karyotype plot (among other things). For a long time our solution has been (you guessed it) ggplot. As powerful as it is, it can get very complicated and intimidating with layers on top of layers!</p>
<p>There comes <a href="https://bernatgel.github.io/karyoploter_tutorial/">karyoploteR</a>!</p>
<p>KaryoploteR is an R package developed by <a href="https://github.com/bernatgel/karyoploteR">bernatgel</a>. It’s a great tool for data visualization in association with chromosomes. It can plot any genomic features (genes, promoters, smRNAs, expression levels, etc) along chromosomes. For this post, I will be focusing on a small problem which only utilizes a tiny fraction of what karyoploteR is capable of. You should definitely check out their full tutorial to harvest its power!</p>
<div id="lets-try-it" class="section level2">
<h2>#Let’s try it</h2>
<p>title: “2-13-19-karyotyploteR” author: “Sichong Peng” date: “February 14, 2019” output: pdf_document: default html_notebook: default —</p>
<p>Prep library</p>
<pre class="r"><code>library(karyoploteR, quietly = TRUE)</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind,
##     colMeans, colnames, colSums, dirname, do.call, duplicated,
##     eval, evalq, Filter, Find, get, grep, grepl, intersect,
##     is.unsorted, lapply, lengths, Map, mapply, match, mget, order,
##     paste, pmax, pmax.int, pmin, pmin.int, Position, rank, rbind,
##     Reduce, rowMeans, rownames, rowSums, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which, which.max,
##     which.min</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## 
## Attaching package: &#39;IRanges&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:grDevices&#39;:
## 
##     windows</code></pre>
<pre class="r"><code>library(readr, quietly = TRUE)
library(dplyr, quietly = TRUE)</code></pre>
<pre><code>## 
## Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:GenomicRanges&#39;:
## 
##     intersect, setdiff, union</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomeInfoDb&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following objects are masked from &#39;package:IRanges&#39;:
## 
##     collapse, desc, intersect, setdiff, slice, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:S4Vectors&#39;:
## 
##     first, intersect, rename, setdiff, setequal, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:BiocGenerics&#39;:
## 
##     combine, intersect, setdiff, union</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     filter, lag</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     intersect, setdiff, setequal, union</code></pre>
<pre class="r"><code>library(ggplot2, quietly = TRUE)
library(qqman,quietly = TRUE)</code></pre>
<pre><code>## </code></pre>
<pre><code>## For example usage please run: vignette(&#39;qqman&#39;)</code></pre>
<pre><code>## </code></pre>
<pre><code>## Citation appreciated but not required:</code></pre>
<pre><code>## Turner, S.D. qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. biorXiv DOI: 10.1101/005165 (2014).</code></pre>
<pre><code>## </code></pre>
<pre class="r"><code>library(stringr, quietly = TRUE)</code></pre>
</div>
<div id="load-data" class="section level2">
<h2>Load data:</h2>
<pre class="r"><code>data = read_tsv(&quot;cc_data.tsv&quot;)</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   CHROM = col_character(),
##   POS = col_double(),
##   CC_ALL = col_double()
## )</code></pre>
<pre class="r"><code>chrom_data = read_tsv(&quot;ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/002/863/925/GCF_002863925.1_EquCab3.0/GCF_002863925.1_EquCab3.0_assembly_report.txt&quot;, comment = &quot;#&quot;, col_names = FALSE)</code></pre>
<pre><code>## Parsed with column specification:
## cols(
##   X1 = col_character(),
##   X2 = col_character(),
##   X3 = col_character(),
##   X4 = col_character(),
##   X5 = col_character(),
##   X6 = col_character(),
##   X7 = col_character(),
##   X8 = col_character(),
##   X9 = col_double(),
##   X10 = col_character()
## )</code></pre>
<pre class="r"><code>chrom_data = filter(chrom_data, X3 != &quot;na&quot;)
chrom_data$start = rep(0,33)
chrom_data_1 = select(chrom_data, X1, start,X9)
colnames(chrom_data_1) = c(&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;)
colnames(data) = c(&quot;CHR&quot;,&quot;BP&quot;,&quot;P&quot;)</code></pre>
</div>
<div id="make-a-custom-equine-genome" class="section level2">
<h2>Make a custom equine genome:</h2>
<pre class="r"><code>genome = toGRanges(data.frame(chrom_data_1))
kp &lt;- plotKaryotype(genome = genome)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
</div>
<div id="correct-chrom-names-in-data" class="section level2">
<h2>Correct chrom names in data</h2>
<pre class="r"><code>data$CHR &lt;- chrom_data$X1[match(data$CHR, chrom_data$X5)]
#Drop those not on chromosome-level assembly
data = filter(data, !is.na(CHR))
data$start = data$BP
data = select(data, CHR, start, BP, P)</code></pre>
</div>
<div id="lets-try-to-plot-them-as-markers-on-karyotypes" class="section level2">
<h2>Let’s try to plot them as markers on karyotypes</h2>
<pre class="r"><code>kp &lt;- plotKaryotype(genome = genome)
kpPlotRegions(kp, toGRanges(data.frame(data)), col = &quot;red&quot;, data.panel=&quot;ideogram&quot;)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div id="so-its-useless" class="section level3">
<h3>So… It’s useless</h3>
</div>
</div>
<div id="what-if-we-try-plotting-those-of-p0.01" class="section level2">
<h2>What if we try plotting those of p&lt;0.01?</h2>
<pre class="r"><code>kp &lt;- plotKaryotype(genome = genome)
kpPlotRegions(kp, toGRanges(data.frame(filter(data, P &lt;= 0.01))), col = &quot;red&quot;, data.panel=&quot;ideogram&quot;)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<div id="its-still-everywhere" class="section level3">
<h3>It’s still everywhere…</h3>
</div>
<div id="now-lets-see-what-p-values-we-have-in-our-data" class="section level3">
<h3>Now let’s see what p-values we have in our data</h3>
<pre class="r"><code>summary(data$P)</code></pre>
<pre><code>##      Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
## 1.000e-08 1.127e-02 2.005e-02 2.384e-02 3.450e-02 4.984e-02</code></pre>
</div>
<div id="histogram" class="section level3">
<h3>Histogram:</h3>
<pre class="r"><code>ggplot(data = data, aes(x = P)) + geom_histogram(binwidth = 0.001)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-8-1.png" width="672" /></p>
</div>
<div id="notice-all-those-missing-p-values-this-is-because-how-few-samples-we-have-here" class="section level3">
<h3>Notice all those missing p-values? This is because how few samples we have here</h3>
</div>
<div id="indeed-in-the-vcf-file-we-have-this-comment-line" class="section level3">
<h3>Indeed, in the vcf file we have this comment line:</h3>
<blockquote>
<p>SnpSiftCmd=“SnpSift CaseControl -f all_var.vcf 0+-++—++——–”</p>
</blockquote>
</div>
<div id="this-tells-us-that-in-this-dataset-there-are-5-cases-and-12-controls.-how-many-different-p-values-can-there-possibly-be-hint-how-many-different-contingency-tables-can-you-draw" class="section level3">
<h3>This tells us that in this dataset there are 5 cases and 12 controls. How many different p-values can there possibly be? (hint: how many different contingency tables can you draw?)</h3>
</div>
<div id="this-is-also-why-a-manhattan-plot-is-often-not-a-good-idea-for-whole-genome-association-studies-unless-you-have-a-lot-more-samples" class="section level3">
<h3>This is also why a Manhattan plot is often not a good idea for whole genome association studies (unless you have a lot more samples)</h3>
</div>
<div id="lets-try-to-plot-a-manhattan-plot" class="section level3">
<h3>Let’s try to plot a Manhattan plot:</h3>
<pre class="r"><code>data_man = filter(mutate(data, CHR = str_extract(CHR, &quot;[1-9]+$&quot;)), !is.na(CHR))
data_man$CHR = as.numeric(data_man$CHR)
manhattan(data_man)</code></pre>
<pre><code>## Warning in manhattan(data_man): No SNP column found. OK unless you&#39;re
## trying to highlight.</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="we-can-also-use-karyoyplotr-to-draw-a-manhattan-plot" class="section level3">
<h3>We can also use karyoyplotR to draw a Manhattan plot:</h3>
<pre class="r"><code>genome_man = toGRanges(data.frame(filter(mutate(chrom_data_1, chr = str_extract(chr, &quot;[1-9]+$&quot;)), !is.na(chr))))
kp &lt;- plotKaryotype(genome = genome_man, plot.type = 4)
kpPoints(kp, data = toGRanges(data.frame(data_man[data_man[,1]%%2==1,])), y = -log10(data_man[data_man[,1]%%2==1,]$P), col=&quot;gold&quot;, ymin = 0, ymax = 10)
kpPoints(kp, data = toGRanges(data.frame(data_man[data_man[,1]%%2==0,])), y = -log10(data_man[data_man[,1]%%2==0,]$P), col=&quot;blue&quot;, ymin = 0, ymax = 10)
kpAxis(kp, ymax = 10)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
</div>
<div id="lets-see-some-more-stringgent-p-values" class="section level3">
<h3>Let’s see some more stringgent p-values:</h3>
<pre class="r"><code>kp &lt;- plotKaryotype(genome = genome)
kpPlotRegions(kp, toGRanges(data.frame(filter(data, P &lt;= 0.0001))), col = &quot;red&quot;, data.panel=&quot;ideogram&quot;)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
</div>
<div id="now-lets-focus-on-chromsome-24" class="section level3">
<h3>Now let’s focus on chromsome 24</h3>
<pre class="r"><code>data_24 = filter(data, CHR == &quot;chr24&quot;)
chrom_24 = toGRanges(data.frame(filter(chrom_data_1, chr == &quot;chr24&quot;)))
data_24$logP = -log10(data_24$P)</code></pre>
</div>
<div id="first-lets-see-variants" class="section level3">
<h3>First let’s see variants</h3>
<pre class="r"><code>kp &lt;- plotKaryotype(genome = chrom_24)
kpPlotRegions(kp, toGRanges(data.frame(data_24)), col = &quot;red&quot;, data.panel=&quot;ideogram&quot;)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
</div>
<div id="this-tells-us-that-a-majority-of-variants-are-in-the-middle-of-chromosome-24" class="section level3">
<h3>This tells us that a majority of variants are in the middle of chromosome 24</h3>
<pre class="r"><code>kp &lt;- plotKaryotype(genome = chrom_24)
kpPoints(kp, data = toGRanges(data.frame(data_24)), y = data_24$logP, col=&quot;red&quot;, ymin = 0, ymax = 8)
kpAxis(kp, ymin = 0, ymax = 8)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<pre class="r"><code>#data.panel=&quot;ideogram&quot;</code></pre>
</div>
<div id="now-we-see-a-chunk-of-them-with-low-p-value-on-chr24.-this-is-what-we-are-looking-for.-lets-get-the-coordinates-of-this-region." class="section level3">
<h3>Now we see a chunk of them with low p-value on chr24. This is what we are looking for. Let’s get the coordinates of this region.</h3>
<pre class="r"><code>data_24_selected = filter(data_24, logP&gt;=4)
summary(data_24_selected$BP)</code></pre>
<pre><code>##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
## 20526862 22673731 24689436 24314147 26029289 27265538</code></pre>
</div>
<div id="now-the-min-and-max-points-to-our-region-chr2420526862-27265538" class="section level3">
<h3>Now the min and max points to our region: chr24:20526862-27265538</h3>
</div>
<div id="lets-zoom-in" class="section level3">
<h3>Let’s zoom in</h3>
<pre class="r"><code>zoom.region &lt;- toGRanges(data.frame(&quot;chr24&quot;, 20526862, 27265538))
kp &lt;- plotKaryotype(genome = chrom_24, zoom = zoom.region)
kpPoints(kp, data = toGRanges(data.frame(data_24)), y = data_24$logP, col=&quot;red&quot;, ymin = 0, ymax = 8)
kpAxis(kp, ymin = 0, ymax = 8)</code></pre>
<p><img src="/post/2019-02-13-draw-karyotype-plot-on-any-organism_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
</div>
</div>
